### DISCUSSION TITLE
Automations with Sunshine: Is there a service that runs when a streaming session is initiated and terminates when it ends?

### DISCUSSION THREAD

**Discussion by AuThomasPrime** [LizardByte/community#209] [Sunshine]
I like to run automations using Home Assistant with Sunshine/Moonlight when a stream is initiated or terminated. For example, I'll automatically change the output on my HDMI switch to a dummy plug, change the desktop resolution to match my client, and turn off my TV when I start a stream. The reverse will happen when I end the stream. It essentially gives my gaming PC "docking" functionality like a Nintendo Switch, but wireless.

Anyway, this requires a reliable indication of when a stream starts and ends. With NVIDIA Gamestream it was simple, because the Gamestream service/exe would only start when a stream was initiated, and stop when the stream ended (which also worked during a crash). Therefore, I could just watch for this service in my automations. However, the Sunshine service seems to run constantly in the background (which has it's own benefits like connection speed), which means I can't tell when a client connects or disconnects directly.

I currently have it working with a dummy service that runs before and stops after a stream via Command Preparations. However, this can be quite buggy when something unexpected happens like a crash, which will leave the dummy service running (because the "Undo" command won't run) and cause the prep-commands to fail on future stream attempts. **Before I spend time programming around this with Batch scripting, I was wondering if there was a more direct method of doing this with Sunshine?**

**Comment by ReenigneArcher** [comment]
The do/undo commands are the best we have right now. We've been thinking of the idea of webhooks or something along those lines, but no work has been done on this as of yet as it's not a big priority.

**Comment by quonic** [comment]
This might help. I have two scripts `Enable-Light.ps1` and `Disable-Light.ps1` .

In sunshine settings, edit Desktop or what ever application, look for `Command Preparations` and click on the `+`.

Add something like this for Do Command:

`powershell.exe -NonInteractive -NoProfile -NoLogo -WindowStyle Hidden -ExecutionPolicy Bypass -File C:\Scripts\Enable-Light.ps1`

Add something like this for Undo Command:

`powershell.exe -NonInteractive -NoProfile -NoLogo -WindowStyle Hidden -ExecutionPolicy Bypass -File C:\Scripts\Disable-Light.ps1`

Enable-Light.ps1:
```powershell
$BaseUrl = "http://homeassistant.local:8123"
$AccessKey = "<Your Token Here>"
$Header = @{ Authorization = "Bearer $AccessKey"; "Content-Type" = "application/json" }
$Entity = "switch.<Some Switch Id Here>"
$On = "/api/services/switch/turn_on"
$Body = @{entity_id = $Entity } | ConvertTo-Json
$uri = "$BaseUrl$On"
Invoke-RestMethod -Method Post -Uri $uri -Headers $Header -Body $Body
```

Disable-Light.ps1:
```powershell
$BaseUrl = "http://homeassistant.local:8123"
$AccessKey = "<Your Token Here>"
$Header = @{ Authorization = "Bearer $AccessKey"; "Content-Type" = "application/json" }
$Entity = "switch.<Some Switch Id Here>"
$Off = "/api/services/switch/turn_off"
$Body = @{entity_id = $Entity } | ConvertTo-Json
$uri = "$BaseUrl$Off"
Invoke-RestMethod -Method Post -Uri $uri -Headers $Header -Body $Body
```

**Comment by JPersson77** [comment]
Hi, any progress with webhooks/ipc/events or other means of discovering when sunshine is streaming / not streaming?

**Reply by ReenigneArcher** [reply]
No, but I feel that this could be integrated with the logger. Meaning whenever we log an event, automatically send a webhook. I have a few things I'm trying to wrap up, but will look into this soon.

**Reply by JPersson77** [reply]
Thanks for your reply. I've built something around parsing the log but unfortunately the CLIENT_DISCONNECT is not guaranteed so it would be nice with something more deterministic :)
